app/services/finder/ByParentEnterpriseUnitFinder.scala

```diff
@@ -0,0 +1,26 @@
+package services.finder
+import com.typesafe.scalalogging.LazyLogging
+import play.api.libs.json.JsObject
+import services.ErrorMessage
+import services.finder.ByParentEnterpriseUnitFinder.UnitRetrieval
+import uk.gov.ons.sbr.models.{ Ern, Period, UnitLinks }
+import unitref.UnitRef
+
+import scala.concurrent.Future
+
+class ByParentEnterpriseUnitFinder[T](retrieveUnit: UnitRetrieval[T], enterpriseUnitRefType: UnitRef[Ern]) extends UnitFinder[T] with LazyLogging {
+  override def find(period: Period, unitRef: T, unitLinks: UnitLinks): Future[Either[ErrorMessage, Option[JsObject]]] =
+    UnitLinks.parentErnFrom(enterpriseUnitRefType)(unitLinks).fold(Future.successful(onMissingParentEnterprise(unitRef))) { ern =>
+      logger.debug(s"Attempting to retrieve unit with [$ern] and [$unitRef] for [$period] ...")
+      retrieveUnit(period, ern, unitRef)
+    }
+
+  private def onMissingParentEnterprise(unitRef: T): Either[ErrorMessage, Option[JsObject]] = {
+    logger.warn(s"Invalid data.  The unit [$unitRef] has Unit Links that do not contain the mandatory parent Enterprise.")
```

Comment:
We are returning a Left here to indicate an error, and all Lefts will be logged at "error" level once they make it to the top-level (the RetrieveLinkedUnitAction).  However, for something of this nature it seems a good idea to have something logged as close to the call site as possible.  I chose warn to avoid one issue triggering multiple errors throughout the logs - which can be a particular pain point once monitoring / alerting is in place.